### PR TITLE
📦 [Dependencies.swift] Decode SPM package information

### PR DESCRIPTION
**Pull request by danyf90**
Part of https://github.com/tuist/tuist/issues/3002

### Short description 📝

Enhance the `SwiftPackageManagerControlling` with the logic to decode SPM package information.

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+        switch key {
+        case .target:
+            var unkeyedValues = try values.nestedUnkeyedContainer(forKey: key)
+            self = .target(
+                name: try unkeyedValues.decode(String.self),
+                condition: try unkeyedValues.decodeIfPresent(PackageConditionDescription.self)
+            )
+        case .product:
+            var unkeyedValues = try values.nestedUnkeyedContainer(forKey: key)
```
Comment (fortmarek) [inline Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift:235]: this line is duplicated but afaict can be done once before `switch key`

File: Sources/TuistDependencies/SwiftPackageManager/Utils/SwiftPackageManagerController.swift [hunk 2]
```diff
@@ -17,6 +17,10 @@ public protocol SwiftPackageManagerControlling {
     /// - Parameter path: Directory where the `Package.swift` is defined.
     /// - Parameter version: Version of tools. When `nil` then the environment’s version will be set.
     func setToolsVersion(at path: AbsolutePath, to version: String?) throws
+
+    /// Loads the information from of the package.
```
Comment (fortmarek) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/SwiftPackageManagerController.swift:21]:
```suggestion
    /// Loads the information from the package.
```

File: Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+/// The Swift Package Manager package information.
+/// It decodes data encoded from Manifest.swift: https://github.com/apple/swift-package-manager/blob/06f9b30f4593940272f57f6284e5614d817d2f22/Sources/PackageModel/Manifest.swift#L372-L409
+/// Fields not needed by tuist are commented out and not decoded at all.
+public struct PackageInfo: Decodable, Equatable {
+    /// The declared platforms in the manifest.
+    public let platforms: [Platform]
+
+    /// The targets declared in the manifest.
+    public let targets: [Target]
```
[the remark this answers is not in the document]
Comment (danyf90) [reply Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift:11]: Probably yes, I'll add it when implementing the mapping, if needed

File: Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    // public let cLanguageStandard: String?
+
+    // /// The C++ language standard flag.
+    // public let cxxLanguageStandard: String?
+
+    // /// The supported Swift language versions of the package.
+    // public let swiftLanguageVersions: [SwiftLanguageVersion]?
+
+    // /// The declared package dependencies.
+    // public let dependencies: [PackageDependencyDescription]
```
[the remark this answers is not in the document]
Comment (danyf90) [reply Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift:39]: Probably yes, I'll add it when implementing the mapping, if needed

**Review by fortmarek — APPROVED** [review]
Apart from the comments about what we might need (which can be implemented later), this is probably good to go from my side 👍

File: Tests/TuistDependenciesTests/SwiftPackageManager/Utils/SwiftPackageManagerControllerTests.swift [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        // When
+        let packageInfo = try subject.loadPackageInfo(at: path)
+
+        // Then
+        XCTAssertEqual(packageInfo, Self.testPackageInfo)
+    }
+}
+
+extension SwiftPackageManagerControllerTests {
+    static let testPackageInfoJSON = """
```
Comment (laxmorek) [inline Tests/TuistDependenciesTests/SwiftPackageManager/Utils/SwiftPackageManagerControllerTests.swift:112]: Could you move it to `TuistDependenicesTesting` framework. 

[Carthage's](https://github.com/tuist/tuist/blob/bb5534bc4def00e7ff28f8e8d1e294da158c07b6/Sources/TuistDependenciesTesting/Carthage/Models/CarthageVersionFile%2BTestData.swift) testing data.

File: Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift [hunk 6]
```diff
@@ -7,11 +7,21 @@ public struct SwiftPackageManagerDependencies: Codable, Equatable {
...
+    /// Creates `SwiftPackageManagerDependencies` instance.
+    /// - Parameter packages: List of packages that will be installed using Swift Package Manager.
+    public init(_ packages: [Package]) {
+        self.packages = packages
+    }
+}
+
+// MARK: - ExpressibleByArrayLiteral
+
+extension SwiftPackageManagerDependencies: ExpressibleByArrayLiteral {
```
Comment (laxmorek) [inline Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift:23]: I added it to my [PR](https://github.com/tuist/tuist/pull/3043).

**Review by laxmorek — APPROVED** [review]
Cool 🎉

**Comment by laxmorek** [comment]
If you want to satisfy `Format` checks you can call `./fourier format swift --fix` in project's root folder.

**Review by FEC-bendingspoons — APPROVED** [review]

**Pull request merged**